### PR TITLE
compose: Add types to `withSafeTimeout`

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Adds types to `withSafeTimeout`. No runtime changes are necessary except that I merge the props outside of the component so that a case to `TProps` is possible. There's effectively no runtime change there though.

Part of #18838

## How has this been tested?
Type checks pass.

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/compose/src/higher-order/with-safe-timeout/index.tsx [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+const withSafeTimeout: PropInjectingHigherOrderComponent< TimeoutProps > = createHigherOrderComponent(
+	< TProps extends TimeoutProps >(
+		OriginalComponent: ComponentType< TProps >
+	) => {
+		return class WrappedComponent extends Component<
+			Omit< TProps, keyof TimeoutProps >
+		> {
+			timeouts: number[];
+
+			constructor( props ) {
```
Comment (ciampo) [inline packages/compose/src/higher-order/with-safe-timeout/index.tsx]: TypeScript doesn't seem to like this line: `Parameter 'props' implicitly has an 'any' type`

Is this enough?

```suggestion
			constructor( props: any ) {
```

File: packages/compose/src/higher-order/with-safe-timeout/index.tsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import createHigherOrderComponent from '../../utils/create-higher-order-component';
+// eslint-disable-next-line no-duplicate-imports
+import type { PropInjectingHigherOrderComponent } from '../../utils/create-higher-order-component';
```
Comment (ciampo) [inline packages/compose/src/higher-order/with-safe-timeout/index.tsx:18]: Is there a reason why we don't use only one import statement?

```suggestion
import createHigherOrderComponent, {
	PropInjectingHigherOrderComponent,
} from '../../utils/create-higher-order-component';
```

Comment (sarayourfriend) [reply packages/compose/src/higher-order/with-safe-timeout/index.tsx:18]: `PropInjectingHigherOrderComponent` is a type and only ever used as a type, so it must be imported using `import type`.

**Review by ciampo — APPROVED** [review]
Changes LGTM

I tried rebasing after noticing some snapshot errors, let's see how it goes

**Pull request merged**